### PR TITLE
pane: Apply `max_tabs` change immediately

### LINKED ISSUE #32217: Setting `max_tabs` should apply the change immediately
### Summary

Currently, changing the `max_tabs` setting does not close tabs out until you trigger the closing when opening a new item. We might want to match other editors and apply it immediately.

### PR DESCRIPTION
**Pull request by vipexv**
Closes #32217

Follow up of https://github.com/zed-industries/zed/pull/32301, sorry about the messy rebase in the previous PR.

Release Notes: 
- Fixed `max_tabs` setting not applying immediately when changed
 
TODO: 
- [x] Fix the off-by-one bug (currently closing one more tab than the max_tabs setting) while perserving "+1 Tab Allowance" feature.
- [x] Investigate Double Invocation of `settings_changed`
- [x] Write test that:
  - Sets max_tabs to `n`
  - Opens `n` buffers
  - Changes max_tabs to `n-1`
  - Asserts we have exactly `n-1` buffers remaining

### CODE REVIEW

**Comment by vipexv** [comment]
@JosephTLyons The latest commit should fix the one off bug, mimicking the behavior of VSCode.

https://github.com/user-attachments/assets/5958af15-c38a-4341-a875-51ee46347005

**Comment by vipexv** [comment]
Marked as draft once again, some tests need to be fixed due to the new changes.

**Comment by vipexv** [comment]
~~During development, I've adjusted the `max_tabs` behavior. Previously, the system allowed `max_tabs + 1` tabs, but my change now strictly enforces `max_tabs`. This has caused some tests to fail because they relied on the old `max_tabs + 1` allowance, do you want me to allow + 1 or just change those tests, we could also add some new logic for specific edge cases.~~

Edit: I tested the old behavior, you already have logic to allow for an extra tab if they're all pinned or other cases, it seems my changes have brought that issue, i'll be working on a fix, any additional info would be appreciated :]

**Comment by vipexv** [comment]
After investigating it some more, i found the root issue for the off-one bug when changing the `max_tabs`, somehow the `observe_global_in` seems to invoke `settings_changed` twice eachtime the settings change leading to a race condition, and i'm unsure why.

**Comment by JosephTLyons** [comment]
Yes, the behavior to allow an extra tab to always be opened, even if we have `max_tabs` or more dirty (unsaved) or pinned tabs is newer and was added so that `max_tabs` never prevents you from opening a file. So you can have `max_tabs` set to 3 and if you have 3 pinned tabs, or 3 unsaved files, or some combination adding up to 3, you can still open a 4th. And if you pin that forth or change it without saving, you can open a 5th, and so on. So we do want to keep that around so that we don't ever block the user from being able to open a file when they've already reached the limit of `max_tabs`.

**Comment by vipexv** [comment]
> Yes, the behavior to allow an extra tab to always be opened, even if we have `max_tabs` or more dirty (unsaved) or pinned tabs is newer and was added so that `max_tabs` never prevents you from opening a file. So you can have `max_tabs` set to 3 and if you have 3 pinned tabs, or 3 unsaved files, or some combination adding up to 3, you can still open a 4th. And if you pin that forth or change it without saving, you can open a 5th, and so on. So we do want to keep that around so that we don't ever block the user from being able to open a file when they've already reached the limit of `max_tabs`.

Yeah i guessed, i went ahead and checked how VSCode does it and they do the same thing, thank you, also if you have any sort of idea as to why `observe_global_in` is invoking `settings_changed` twice eachtime the settings change please let me know :]

**Comment by JosephTLyons** [comment]
> After investigating it some more, i found the root issue for the off-one bug when changing the `max_tabs`, somehow the `observe_global_in` seems to invoke `settings_changed` twice eachtime the settings change leading to a race condition, and i'm unsure why.

Ahh, that would make sense then as to why we're seeing it kill off too many tabs. That being said, that bit might be too deep in the weeds for me to know of a proper fix and I might need to call on another engineer who knows that area more.

**Comment by vipexv** [comment]
> > After investigating it some more, i found the root issue for the off-one bug when changing the `max_tabs`, somehow the `observe_global_in` seems to invoke `settings_changed` twice eachtime the settings change leading to a race condition, and i'm unsure why.
> 
> Ahh, that would make sense then as to why we're seeing it kill off too many tabs. That being said, that bit might be too deep in the weeds for me to know of a proper fix and I might need to call on another engineer who knows that area more.

I have a potential workaround for the double-invocation issue, though it's not ideal:

**Option 1: Context-aware logic**
Add a parameter to the tab-closing function that indicates whether it's being called from `settings_changed`. This would let us handle the double-invocation scenario differently.

**Option 2: Adjust the comparison operator**
Change `items_len < max_tabs` to `items_len <= max_tabs`. This fixes the immediate issue, but it breaks the "+1 tab allowance" feature for pinned/dirty tabs that you mentioned.

Both approaches feel like workarounds rather than proper fixes. The root cause is still the double invocation from `observe_global_in`, which ideally should be addressed at the observer level.

Would it be worth investigating why the observer fires twice? Or should we implement one of these workarounds as a temporary solution while the underlying issue gets proper attention from someone more familiar with the observer system?

**Comment by JosephTLyons** [comment]
Having more context as to why `settings_changed` is called twice would be good. I'm curious if there's a real reason for that or if it's a bug. I've asked the team, and I can also try to dig into it a bit myself this week. If it's an unintentional thing, getting that fixed would be awesome. If its intentional and other things rely on it, a workaround could be added, but I'd prefer to avoid losing the +1 tab behavior, as I think blocking a tab when a user needs it is worse than the bug of us not applying `max_tabs` immediately on settings file save.

**Comment by vipexv** [comment]
> Having more context as to why `settings_changed` is called twice would be good. I'm curious if there's a real reason for that or if it's a bug. I've asked the team, and I can also try to dig into it a bit myself this week. If it's an unintentional thing, getting that fixed would be awesome. If its intentional and other things rely on it, a workaround could be added, but I'd prefer to avoid losing the +1 tab behavior, as I think blocking a tab when a user needs it is worse than the bug of us not applying `max_tabs` immediately on settings file save.

I'll try and look into it myself when i get the chance, thank you once again!

**Comment by JosephTLyons** [comment]
@smitbarmase just hopped into a call with me to show me an existing pattern that might work as a workaround here.

In this code, we observe the settings, but do not run actions on settings change unless the setting value is different from one that was seen last

https://github.com/zed-industries/zed/blob/a9d0eee2a9687c14c5a82d801a185ec280f05cf3/crates/project_panel/src/project_panel.rs#L460-L475

We could do something similar here where we store the `max_tabs` value in the `pane`. 

When observing, we can check if the new setting differs from what we have cached, and avoid double calling `self.close_items_over_max_tabs`.

**Comment by vipexv** [comment]
> @smitbarmase just hopped into a call with me to show me an existing pattern that might work as a workaround here.
> 
> In this code, we observe the settings, but do not run actions on settings change unless the setting value is different from one that was seen last
> 
> https://github.com/zed-industries/zed/blob/a9d0eee2a9687c14c5a82d801a185ec280f05cf3/crates/project_panel/src/project_panel.rs#L460-L475
> 
> We could do something similar here where we store the `max_tabs` value in the `pane`.
> 
> When observing, we can check if the new setting differs from what we have cached, and avoid double calling `self.close_items_over_max_tabs`.

I'm already doing that, but i think i wrote the check wrong, i'll be attempting a fix.

**Comment by vipexv** [comment]
Yeah i don't think my check is wrong.

```rust

        if new_max_tabs != self.current_max_tabs {
            self.current_max_tabs = new_max_tabs;
            self.close_items_over_max_tabs(window, cx);
        }
```
I added some logs and this is what it shows.

image


```rust
        println!(
            "new_max_tabs: - {:?}, current_max_tabs - {:?}, 1 - {:?}",
            new_max_tabs,
            self.current_max_tabs,
            new_max_tabs != self.current_max_tabs
        );
```

**Comment by vipexv** [comment]
> @smitbarmase just hopped into a call with me to show me an existing pattern that might work as a workaround here.
> 
> In this code, we observe the settings, but do not run actions on settings change unless the setting value is different from one that was seen last
> 
> https://github.com/zed-industries/zed/blob/a9d0eee2a9687c14c5a82d801a185ec280f05cf3/crates/project_panel/src/project_panel.rs#L460-L475
> 
> We could do something similar here where we store the `max_tabs` value in the `pane`.
> 
> When observing, we can check if the new setting differs from what we have cached, and avoid double calling `self.close_items_over_max_tabs`.

I just tested the project panel. Interestingly, theirs doesn’t trigger twice.

**Comment by JosephTLyons** [comment]
Interesting! Somehow, a terminal tab is being included as the additional item

```rs
        let item_tabs: Vec<String> = self
            .items
            .iter()
            .map(|item| item.tab_content_text(0, cx).to_string())
            .collect();

        dbg!("Pane::settings_changed", cx.entity_id(), &item_tabs);
```

```
[crates/workspace/src/pane.rs:648:9] "Pane::settings_changed" = "Pane::settings_changed"
[crates/workspace/src/pane.rs:648:9] cx.entity_id() = EntityId(
    79v1,
)
[crates/workspace/src/pane.rs:648:9] &item_tabs = [
    "~/.config/zed/settings.json",
]
[crates/workspace/src/pane.rs:648:9] "Pane::settings_changed" = "Pane::settings_changed"
[crates/workspace/src/pane.rs:648:9] cx.entity_id() = EntityId(
    188v1,
)
[crates/workspace/src/pane.rs:648:9] &item_tabs = [
    "josephlyons — zsh ",
]
```

**Comment by vipexv** [comment]
> Interesting! Somehow, a terminal tab is being included as the additional item
> 
> 
> 
> ```rs
> 
>         let item_tabs: Vec<String> = self
> 
>             .items
> 
>             .iter()
> 
>             .map(|item| item.tab_content_text(0, cx).to_string())
> 
>             .collect();
> 
> 
> 
>         dbg!("Pane::settings_changed", cx.entity_id(), &item_tabs);
> 
> ```
> 
> 
> 
> ```
> 
> [crates/workspace/src/pane.rs:648:9] "Pane::settings_changed" = "Pane::settings_changed"
> 
> [crates/workspace/src/pane.rs:648:9] cx.entity_id() = EntityId(
> 
>     79v1,
> 
> )
> 
> [crates/workspace/src/pane.rs:648:9] &item_tabs = [
> 
>     "~/.config/zed/settings.json",
> 
> ]
> 
> [crates/workspace/src/pane.rs:648:9] "Pane::settings_changed" = "Pane::settings_changed"
> 
> [crates/workspace/src/pane.rs:648:9] cx.entity_id() = EntityId(
> 
>     188v1,
> 
> )
> 
> [crates/workspace/src/pane.rs:648:9] &item_tabs = [
> 
>     "josephlyons — zsh ",
> 
> ]
> 
> ```

I'll be able to investigate this some more when I'm fully back, which will most likely be Thursday.

File: crates/workspace/src/pane.rs [hunk 1]
```diff
@@ -1409,29 +1422,59 @@ impl Pane {
...
+        self.close_tabs_to_target(target, protect_active_item, window, cx);
+    }
+
+    fn close_tabs_to_target(
+        &mut self,
+        target_count: Option<usize>,
+        protect_active_item: bool,
+        window: &mut Window,
+        cx: &mut Context<Self>,
+    ) {
```
Comment (JosephTLyons) [inline crates/workspace/src/pane.rs:1444]: This is essentially the same option that you had mentioned previously, with tweaking how items get closed depending on whether the closing is triggered via a settings change or a file open, with the additional logic to not close the active item in the case of the settings file change.

I don't fully understand the double invocation of the settings change, but I know its caused by the system thinking some terminal tab is open, and this probably isn't the PR to attempt to change whatever is causing that and I dont think the double invocation was causing the failure prior, so even once thats fixed, this code should continue to operate and work as expected.

File: crates/workspace/src/pane.rs [hunk 2]
```diff
@@ -1409,29 +1422,59 @@ impl Pane {
         )
     }
 
-    pub fn close_items_over_max_tabs(&mut self, window: &mut Window, cx: &mut Context<Self>) {
-        let Some(max_tabs) = WorkspaceSettings::get_global(cx).max_tabs.map(|i| i.get()) else {
```
Comment (JosephTLyons) [inline crates/workspace/src/pane.rs:1413]: In the original code, we always pulled from the settings, but now that we are caching the `max_tabs` value on the `pane` struct, I think it makes sense to access the cached value over reaching into settings, which is what I'm now doing in `close_items_on_settings_change` and `close_items_on_file_open`.

**Comment by JosephTLyons** [comment]
Hey @vipexv, I pushed up a commit and added some inline comments and I'm curious what you think. I've also added a bit of extra testing and code to ensure that in the case of closing tabs when the settings file is changed, we dont allow the active file (which is the settings.json file) to be closed, as I think that would likely be a bad experience, so it is now a "protected" item.

Hope me collaborating with you on this was ok! And if anything seems wrong or off about what I pushed, would love to continue chatting about the next steps.

File: crates/workspace/src/pane.rs [hunk 3]
```diff
@@ -1409,29 +1422,59 @@ impl Pane {
...
+    fn close_items_on_file_open(&mut self, window: &mut Window, cx: &mut Context<Self>) {
+        let target = self.max_tabs.map(|m| m.get());
```
Comment (JosephTLyons) [inline crates/workspace/src/pane.rs:1426]: Using a `NonZeroUsize` makes this is a little ugly, as we have to map it to a normal `usize`, but that came from a prior PR and requires tweaking code in our new settings importers (VS Code, etc.), which I know nothing about, so I left it.

**Comment by vipexv** [comment]
> Hey @vipexv, I pushed up a commit and added some inline comments and I'm curious what you think. I've also added a bit of extra testing and code to ensure that in the case of closing tabs when the settings file is changed, we dont allow the active file (which is the settings.json file) to be closed, as I think that would likely be a bad experience, so it is now a "protected" item.
> 
> Hope me collaborating with you on this was ok! And if anything seems wrong or off about what I pushed, would love to continue chatting about the next steps.



> Hey @vipexv, I pushed up a commit and added some inline comments and I'm curious what you think. I've also added a bit of extra testing and code to ensure that in the case of closing tabs when the settings file is changed, we dont allow the active file (which is the settings.json file) to be closed, as I think that would likely be a bad experience, so it is now a "protected" item.
> 
> Hope me collaborating with you on this was ok! And if anything seems wrong or off about what I pushed, would love to continue chatting about the next steps.

Thank you for collaborating with me on this! I also appreciate the comments as it helps me and other people understand why it's written that way :], and i also feel like the settings file should always be protected in this case for a better UX.

I'll be looking into this & testing it some more today, thank you once again for taking the time to work on this!

**Comment by vipexv** [comment]
I marked it as ready for review, the changes from @JosephTLyons have resolved all the issues we had, thank you once again for the effort you've put into fixing this issue :]

**Comment by JosephTLyons** [comment]
Thanks for getting this PR up and going and digging in. I went ahead and merged because it felt good to me. We can always forward fix if something else comes up! Love getting all these tests around all the cases that come up with the various tab states and `max_tab` setting mixtures.

Also, nice work on the other PRs you have landed - these fixes are super appreciated.
